### PR TITLE
Snow 847337 External browser authentication timeout (nodejs)

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,4 +13,4 @@ __pycache__/
 wss-*-agent.config
 wss-unified-agent.jar
 whitesource/
-
+.nyc_output
```
Comment (sfc-gh-dprzybysz) [inline .gitignore:16]: it should not be part of this PR

File: lib/authentication/auth_web.js [hunk 2]
```diff
@@ -9,6 +9,8 @@ var net = require('net');
 var querystring = require('querystring');
 var URLUtil = require('./../../lib/url_util');
 
+const WAIT_FOR_BROWSER_ACTION_TIMEOUT = 120000;
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:12]: it should be part of global config

File: lib/authentication/auth_web.js [hunk 3]
```diff
@@ -19,12 +21,13 @@ var URLUtil = require('./../../lib/url_util');
  * @returns {Object}
  * @constructor
  */
-function auth_web(host, webbrowser, httpclient)
+function auth_web(host, webbrowser, httpclient, browserActionTimeout)
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:24]: we can get it from global config, not as parameter

File: lib/authentication/auth_web.js [hunk 4]
```diff
@@ -19,12 +21,13 @@ var URLUtil = require('./../../lib/url_util');
...
 {
   var open = typeof webbrowser !== "undefined" ? webbrowser : require('open');
   var axios = typeof httpclient !== "undefined" ? httpclient : require('axios');
 
   var host = host;
+  var browserTimeout = browserActionTimeout || WAIT_FOR_BROWSER_ACTION_TIMEOUT
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:30]: should it be let or const?

File: lib/authentication/auth_web.js [hunk 5]
```diff
@@ -58,19 +61,17 @@ function auth_web(host, webbrowser, httpclient)
    */
   this.authenticate = async function (authenticator, serviceName, account, username)
   {
-    var server;
+      var server;
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:64]: when you rewrite - let's introduce let/const

File: lib/authentication/auth_web.js [hunk 6]
```diff
@@ -87,8 +88,8 @@ function auth_web(host, webbrowser, httpclient)
     // Step 3: open browser
     open(ssoURL);
 
-    // Step 4: get SAML token
-    data = await receiveData;
+      // Step 4: get SAML token
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:91]: this comment or line below is not properly indented

File: lib/authentication/auth_web.js [hunk 7]
```diff
@@ -147,7 +148,7 @@ function auth_web(host, webbrowser, httpclient)
   function getSSOURL(authenticator, serviceName, account, callback_port, user)
   {
     // Create URL to send POST request to
-    var url = protocol + '://' + host + "/session/authenticator-request";
+    var url = protocol + '://snowflake.temptest000043.us-west-2.aws.snowflakecomputing.com/session/authenticator-request';
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:151]: let's not use this address here

File: lib/connection/connection_config.js [hunk 8]
```diff
@@ -147,7 +147,7 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
   {
     // username is not required for oauth authenticator
     if (!Util.exists(options.authenticator) ||
-      options.authenticator !== authenticationTypes.OAUTH_AUTHENTICATOR) 
+      options.authenticator !== authenticationTypes.OAUTH_AUTHENTICATOR )
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:150]: if you change formatting then fix also `{}` position but maybe you shouldn't in this PR in this place

File: lib/connection/connection_config.js [hunk 9]
```diff
@@ -260,6 +260,7 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
 
   var serviceName = options.serviceName;
   var authenticator = options.authenticator;
+  var browserActionTimeout = options.browserActionTimeout;
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:263]: let or const?

File: lib/connection/connection_config.js [hunk 10]
```diff
@@ -543,6 +544,16 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
     return authenticator;
   };
 
+  /**
+   * Returns the timeout in millis used for authentication by external browser.
+   *
+   * @returns {String}
+   */
+  this.getBrowserActionTimeout = function ()
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:552]: is it connection or global config option?

File: test/integration/connectionOptions.js [hunk 11]
```diff
@@ -105,6 +105,19 @@ var externalBrowser =
   authenticator: 'EXTERNALBROWSER'
 };
 
+var externalBrowserWithShortTimeout =
```
Comment (sfc-gh-dprzybysz) [inline test/integration/connectionOptions.js:108]: let or const?

File: test/integration/testConnection.js [hunk 12]
```diff
@@ -192,6 +192,26 @@ describe.skip('Connection test - external browser', function ()
...
 
+
+  it('Connect - external browser timeout', function (done) {
+      var connection = snowflake.createConnection(connOption.externalBrowserWithShortTimeout);
+
+      async.series([
+        function (callback) {
+          connection.connectAsync(function (err) {
+            try {
+              let browserActionTimeout = connOption.externalBrowserWithShortTimeout.browserActionTimeout || 120000;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:203]: why do we have default value here? we are not sure we receive it?

File: test/integration/testConnection.js [hunk 13]
```diff
@@ -192,6 +192,26 @@ describe.skip('Connection test - external browser', function ()
     );
   });
 
+
+  it('Connect - external browser timeout', function (done) {
+      var connection = snowflake.createConnection(connOption.externalBrowserWithShortTimeout);
+
+      async.series([
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:199]: we don't need `series` for one action

**Comment by sfc-gh-dszmolka** [comment]
can we please consider having a descriptive name for the PR ?

File: lib/authentication/auth_web.js (hunk 2 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:12]: It isn't set by global config. This timeout is only in connection options and constant default is inside Authenticator class.

File: lib/authentication/auth_web.js (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:24]: It is connection config parameter, doesn't exist in global config

File: lib/authentication/auth_web.js (hunk 4 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:30]: should be const

File: lib/connection/connection_config.js (hunk 9 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection_config.js:263]: even const is possible

File: lib/connection/connection_config.js (hunk 10 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection_config.js:552]: It is connection configuration param

File: test/integration/connectionOptions.js (hunk 11 above)
Comment (sfc-gh-pmotacki) [reply test/integration/connectionOptions.js:108]: const ;)

File: test/integration/testConnection.js (hunk 13 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testConnection.js:199]: That's true

File: .gitignore (hunk 1 above)
Comment (sfc-gh-pmotacki) [reply .gitignore:16]: true

File: lib/authentication/auth_web.js [hunk 14]
```diff
@@ -2,36 +2,45 @@
...
+  }
+  if (!Util.number.isPositiveInteger(browserActionTimeout)) {
+    throw new Error(util.format("Invalid value for browser action timeout ", browserActionTimeout));
+  }
 
-  var successResponse = 'HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nConnection: close\r\n\r\nYour identity was confirmed and propagated to Snowflake Node.js driver. You can close this window now and go back where you started from.';
+  const open = typeof webbrowser !== "undefined" ? webbrowser : require('open');
+  const axios = typeof httpclient !== "undefined" ? httpclient : require('axios');
+
+  const browserTimeout = browserActionTimeout || WAIT_FOR_BROWSER_ACTION_TIMEOUT
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:36]: it's checked above that `browserActionTimeout ` is positive integer - we don't need to set up default

File: lib/connection/connection.js [hunk 15]
```diff
@@ -327,7 +327,7 @@ function Connection(context)
     }
     catch (authErr)
     {
-      throw authErr;
+      callback(authErr)
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection.js:330]: have you tested this branch? it looks like fix but maybe also behaviour change

File: lib/connection/connection_config.js [hunk 16]
```diff
@@ -271,6 +271,15 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
     authenticator = authenticator.toUpperCase();
   }
 
+  let browserActionTimeout = options.browserActionTimeout;
+  if (Util.exists(options.browserActionTimeout))
+  {
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:276]: let's format the code according to eslint rules

File: test/integration/testConnection.js [hunk 17]
```diff
@@ -46,34 +46,34 @@ describe('Connection test', function ()
...
-        function (callback)
-        {
-          connection.connect(function (err)
-          {
-            assert.ok(!err, JSON.stringify(err));
-            callback();
-          });
-        },
-        function (callback)
+      function (callback)
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:49]: if you change this code then let's apply eslint formatting rules

File: test/integration/testConnection.js [hunk 18]
```diff
@@ -46,34 +46,34 @@ describe('Connection test', function ()
...
+      {
+        connection.connect(function (err)
         {
-          assert.ok(connection.isUp(), "not active");
+          assert.ok(!err, JSON.stringify(err));
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:53]: let's try/catch and call callback with error

File: test/integration/testConnection.js [hunk 19]
```diff
@@ -152,62 +152,64 @@ describe('Connection test', function ()
...
-describe.skip('Connection test - external browser', function ()
-{
+describe.skip('Connection test - external browser', function () {
   this.timeout(10000);
 
-  it('Simple Connect', function (done)
-  {
-    var connection = snowflake.createConnection(connOption.externalBrowser);
+  it('Simple Connect', function (done) {
+    let connection = snowflake.createConnection(connOption.externalBrowser);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:159]: should it be const?

File: test/integration/testConnection.js [hunk 20]
```diff
@@ -152,62 +152,64 @@ describe('Connection test', function ()
...
+      },
+    ],
     );
   });
 
-  it('Mismatched Username', function (done)
-  {
+
+  it('Connect - external browser timeout', function (done) {
+      var connection = snowflake.createConnection(connOption.externalBrowserWithShortTimeout);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:189]: let's fix formatting

File: test/integration/testConnection.js [hunk 21]
```diff
@@ -152,62 +152,64 @@ describe('Connection test', function ()
...
+
+    connection.connectAsync(function (err) {
+      try {
+        let browserActionTimeout = connOption.externalBrowserWithShortTimeout.browserActionTimeout || 120000;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:193]: it could be const

File: test/unit/authentication/authentication_test.js [hunk 22]
```diff
@@ -109,7 +109,7 @@ describe('external browser authentication', function ()
 
   it('external browser - authenticate method is thenable', done =>
   {
-    const auth = new auth_web('', webbrowser.open, httpclient);
+    const auth = new auth_web('', webbrowser.open, httpclient, 10000);
```
Comment (sfc-gh-dprzybysz) [inline test/unit/authentication/authentication_test.js:112]: let's move 10000s to named const

File: test/unit/authentication/authentication_test.js [hunk 23]
```diff
@@ -160,7 +160,7 @@ describe('external browser authentication', function ()
     webbrowser = require('webbrowser');
     httpclient = require('httpclient');
 
-    var auth = new auth_web('', webbrowser.open, httpclient);
+    var auth = new auth_web('', webbrowser.open, httpclient, -10000);
```
Comment (sfc-gh-dprzybysz) [inline test/unit/authentication/authentication_test.js:163]: should it crash? the crash reason will be the missing url or negative timeout?

File: test/integration/testConnection.js (hunk 19 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testConnection.js:159]: fixed

File: lib/authentication/auth_web.js (hunk 14 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:36]: ok

File: lib/connection/connection_config.js (hunk 16 above)
Comment (sfc-gh-pmotacki) [reply lib/connection/connection_config.js:276]: sure

File: test/integration/testConnection.js (hunk 17 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testConnection.js:49]: ok

File: test/unit/authentication/authentication_test.js (hunk 23 above)
Comment (sfc-gh-pmotacki) [reply test/unit/authentication/authentication_test.js]: fixed

File: lib/authentication/authentication.js [hunk 24]
```diff
@@ -77,7 +77,7 @@ exports.getAuthenticator = function getAuthenticator(connectionConfig)
   }
   else if (auth == authenticationTypes.EXTERNAL_BROWSER_AUTHENTICATOR)
   {
-    return new auth_web(connectionConfig.host);
+    return new auth_web(connectionConfig.host, undefined, undefined, connectionConfig.getBrowserActionTimeout());
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/authentication.js:80]: in production code second and third parameter are always undefined - let's move them to the end since timeout is obligatory here

File: lib/authentication/auth_web.js [hunk 25]
```diff
@@ -2,36 +2,45 @@
...
-  var axios = typeof httpclient !== "undefined" ? httpclient : require('axios');
+function auth_web(host, webbrowser, httpclient, browserActionTimeout) {
 
-  var host = host;
-  var port = rest.HTTPS_PORT;
-  var protocol = rest.HTTPS_PROTOCOL;
-  var proofKey;
-  var token;
-  var data;
+  if (!Util.exists(host)) {
```
Comment (sfc-gh-dprzybysz) [inline lib/authentication/auth_web.js:26]: do we need those assertions? it's internal code and if host or timeout are missing then connection config will fail fast?

File: lib/connection/connection_config.js [hunk 26]
```diff
@@ -543,6 +551,16 @@ function ConnectionConfig(options, validateCredentials, qaMode, clientInfo)
     return authenticator;
   };
 
+  /**
+   * Returns the timeout in millis used for authentication by external browser.
+   *
+   * @returns {String}
+   */
+  this.getBrowserActionTimeout = function ()
+  {
```
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:560]: let's format new code

File: test/integration/testConnection.js [hunk 27]
```diff
@@ -77,6 +77,39 @@ describe('Connection test', function ()
     );
   });
 
+  // it('Simple Connect2', function (done)
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:80]: we don't need commented code

File: test/integration/testConnection.js [hunk 28]
```diff
@@ -152,62 +185,64 @@ describe('Connection test', function ()
...
-            assert.ok(!err, JSON.stringify(err));
-            callback();
-          });
-        },
-        function (callback)
-        {
-          assert.ok(connection.isUp(), "not active");
+      function (callback) {
+        connection.connectAsync(function (err) {
+          done(err);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:197]: after done you should not do anything more

File: test/integration/testConnection.js [hunk 29]
```diff
@@ -152,62 +185,64 @@ describe('Connection test', function ()
...
-            callback();
-          });
-        },
-        function (callback)
-        {
-          assert.ok(!connection.isUp(), "still active");
+        });
+      },
+      function (callback) {
+        assert.ok(connection.isUp(), 'not active');
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:203]: we should wrap assertions in try/catch and pass an error to callback if happen

File: test/integration/testConnection.js [hunk 30]
```diff
@@ -152,62 +185,64 @@ describe('Connection test', function ()
...
 
-  it('Mismatched Username', function (done)
-  {
+
+  it('Connect - external browser timeout', function (done) {
+    const connection = snowflake.createConnection(connOption.externalBrowserWithShortTimeout);
+
+    connection.connectAsync(function (err) {
+      try {
+        const browserActionTimeout = connOption.externalBrowserWithShortTimeout.browserActionTimeout || 120000;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testConnection.js:226]: we should not have here default value

**Review by sfc-gh-dprzybysz — CHANGES_REQUESTED** [review]

File: test/integration/testConnection.js (hunk 27 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testConnection.js]: of course

File: lib/authentication/auth_web.js [hunk 31]
```diff
@@ -2,36 +2,45 @@
...
+  const axios = typeof httpclient !== "undefined" ? httpclient : require('axios');
+
+  const browserTimeout = browserActionTimeout
+  const port = rest.HTTPS_PORT;
+  const protocol = rest.HTTPS_PROTOCOL;
+  let proofKey;
+  let token;
+  let data;
+
+  let successResponse = 'HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nConnection: close\r\n\r\nYour identity was confirmed and propagated to Snowflake Node.js driver. You can close this window now and go back where you started from.';
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:43]: ok

File: lib/authentication/auth_web.js [hunk 32]
```diff
@@ -158,7 +164,7 @@ function auth_web(host, webbrowser, httpclient)
     }
 
     // JSON body to send with POST request
-    var body = {
+    let body = {
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:167]: ok

File: test/integration/testConnection.js (hunk 30 above)
Comment (sfc-gh-pmotacki) [reply test/integration/testConnection.js]: ok

File: lib/authentication/auth_web.js (hunk 25 above)
Comment (sfc-gh-pmotacki) [reply lib/authentication/auth_web.js:26]: I check to be sure that I have all needed parameters

File: test/integration/connectionOptions.js [hunk 33]
```diff
@@ -105,15 +105,28 @@ var externalBrowser =
   authenticator: 'EXTERNALBROWSER'
 };
 
-var externalBrowserMismatchUser =
+const externalBrowserWithShortTimeout =
```
Comment (sfc-gh-dprzybysz) [inline test/integration/connectionOptions.js:108]: let's create this configuration as 

```
const externalBrowserWithShortTimeout = {
  ...externalBrowser,
  browserActionTimeout: 100,
};
```

to not copy past most of the configuration

**Review by sfc-gh-dprzybysz — APPROVED** [review]

File: test/integration/connectionOptions.js (hunk 33 above)
Comment (sfc-gh-pmotacki) [reply test/integration/connectionOptions.js:108]: You wright Fixed

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM
